edgelet/edgelet-docker/src/runtime.rs

```diff
@@ -350,13 +354,32 @@ impl ModuleRuntime for DockerModuleRuntime {
                     module.config().image()
                 );
 
-                let create_options = create_options
+                let mut create_options = create_options
                     .with_image(module.config().image().to_string())
                     .with_env(merged_env)
                     .with_labels(labels);
 
-                // Here we don't add the container to the iot edge docker network as the edge-agent is expected to do that.
-                // It contains the logic to add a container to the iot edge network only if a network is not already specified.
+                // In the bootstrap scenario (edge runtime module starting up for the first time),
```

Comment:
Why can't we update the spec for the EdgeAgent that the watchdog uses, instead? 
https://github.com/Azure/iotedge/blob/master/edgelet/iotedged/src/lib.rs#L1978
(In short, instead of fixing it in the lower layer, why don't we fix it in the higher layer)?

Reply:
The watchdog uses an opaque generic for the module spec because it's generic on the module runtime, so it can't see it as a `DockerConfig` and mutate it.

Reply:
That's what I originally tried to do. But `DockerModuleRuntime::create` is the first function in that call chain that knows anything about Docker. The network needs to be added to the create options, and I can't do that at the higher level. But the problem with `create` is that it gets called for all modules, not just agent and not just the first time it starts. So I had to come up with a way to discriminate.

There might be a way to do it but I couldn't find one...

Reply:
We could add a new trait for the config, with a `fn set_network(&mut self, network_id: String);` and make it a bound for https://github.com/Azure/iotedge/blob/b1e6ba543f5be7b4c6cb9b6d7d7f2d97b401e203/edgelet/edgelet-core/src/module.rs#L146 . But it would touch a lot more files than what this PR does now, though it *would* definitely be cleaner.

Alternatively we could add the trait but only add the bound for the watchdog function rather than `ModuleSpec` itself. I'm leaning towards this now that I think about it.

Reply:
@damonbarry I suggest:

- Add
    ```rust
    trait Config {
        fn set_network(&mut self, network_id: String);
    }
    ```

    somewhere under `edgelet-core`

- impl it for `edgelet_docker::DockerConfig` and friends

- change https://github.com/Azure/iotedge/blob/b1e6ba543f5be7b4c6cb9b6d7d7f2d97b401e203/edgelet/edgelet-core/src/watchdog.rs#L42 to

    ```rust
    <M::Module as Module>::Config: Clone + Config,
    ```

  (Adding the bound here rather than on `struct ModuleSpec` itself means you don't have to change the rest of the codebase that uses `ModuleSpec`.

- and add a `spec.set_network()` call in https://github.com/Azure/iotedge/blob/b1e6ba543f5be7b4c6cb9b6d7d7f2d97b401e203/edgelet/edgelet-core/src/watchdog.rs#L56 You'll also have to thread `settings.moby_runtime().network()` to the watchdog from main.

Reply:
I'll give it a try. I definitely wasn't loving taking advantage of a side-effect as a solution. Thanks!

Reply:
> You'll also have to thread settings.moby_runtime().network() to the watchdog from main.

@arsing Does that mean I need to add a new trait for Settings too (with, e.g., `fn get_network()`)? `M::Settings` constraints in Main are super generic...not docker-specific.

Reply:
Oh, drat. I forgot about that.

There's already [`RuntimeSettings`](https://github.com/Azure/iotedge/blob/53757e2be8bc9aedab41d46b9472ac7b8cca3cd0/edgelet/edgelet-core/src/settings.rs#L677) for that, but looking at the impls I see that `moby_runtime` is [`edgelet-docker`-specific.](https://github.com/Azure/iotedge/blob/53757e2be8bc9aedab41d46b9472ac7b8cca3cd0/edgelet/edgelet-docker/src/settings.rs#L52) So it would run counter to the purpose of that trait to expose `fn network(&self) -> &str` on it, and not be possible to do for [`edgelet-kube`'s struct](https://github.com/Azure/iotedge/blob/53757e2be8bc9aedab41d46b9472ac7b8cca3cd0/edgelet/edgelet-kube/src/settings.rs#L19) anyway.